File: spark/wallet/transaction.go
package wallet

// Tools for building all the different transactions we use.

import (
	"fmt"

	"github.com/btcsuite/btcd/wire"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/lightsparkdev/spark-go"
	"github.com/lightsparkdev/spark-go/common"
)

func createRootTx(
	depositOutPoint *wire.OutPoint,
	depositTxOut *wire.TxOut,
) *wire.MsgTx {
	rootTx := wire.NewMsgTx(2)
	rootTx.AddTxIn(wire.NewTxIn(depositOutPoint, nil, nil))
	// We currently send the full value to the same address
	// TODO: 0 fee will only be okay once we add ephemeral anchor outputs
	rootTx.AddTxOut(depositTxOut)
	return rootTx
}

func nextSequence(currSequence uint32) (uint32, error) {
	if currSequence&0xFFFF-spark.TimeLockInterval <= 0 {
		return 0, fmt.Errorf("timelock interval is less or equal to 0")
	}
	return uint32((1 << 30) | (currSequence&0xFFFF - spark.TimeLockInterval)), nil
}

func createRefundTx(
	sequence uint32,
	nodeOutPoint *wire.OutPoint,
	amountSats int64,
	receivingPubkey *secp256k1.PublicKey,
) (*wire.MsgTx, error) {
	newRefundTx := wire.NewMsgTx(2)
	newRefundTx.AddTxIn(&wire.TxIn{
		PreviousOutPoint: *nodeOutPoint,
		SignatureScript:  nil,
		Witness:          nil,
		Sequence:         sequence,
	})

	refundPkScript, err := common.P2TRScriptFromPubKey(receivingPubkey)
	if err != nil {
		return nil, fmt.Errorf("failed to create refund pkscript: %v", err)
	}
	newRefundTx.AddTxOut(wire.NewTxOut(amountSats, refundPkScript))

	return newRefundTx, nil
}

func createConnectorRefundTransaction(
	sequence uint32,
	nodeOutPoint *wire.OutPoint,
	connectorOutput *wire.OutPoint,
	amountSats int64,
	receiverPubKey *secp256k1.PublicKey,
) (*wire.MsgTx, error) {
	refundTx := wire.NewMsgTx(2)
	refundTx.AddTxIn(&wire.TxIn{
		PreviousOutPoint: *nodeOutPoint,
		SignatureScript:  nil,
		Witness:          nil,
		Sequence:         sequence,
	})
	refundTx.AddTxIn(wire.NewTxIn(connectorOutput, nil, nil))
	receiverScript, err := common.P2TRScriptFromPubKey(receiverPubKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create receiver script: %v", err)
	}
	refundTx.AddTxOut(wire.NewTxOut(amountSats, receiverScript))
	return refundTx, nil
}

File: sdks/js/packages/spark-sdk/src/utils/transaction.ts
import { Transaction } from "@scure/btc-signer";
import { TransactionInput, TransactionOutput } from "@scure/btc-signer/psbt";
import { getP2TRScriptFromPublicKey } from "./bitcoin.js";
import { Network } from "./network.js";

const TIME_LOCK_INTERVAL = 100;

export function createRefundTx(
  sequence: number,
  nodeOutPoint: TransactionInput,
  amountSats: bigint,
  receivingPubkey: Uint8Array,
  network: Network,
): Transaction {
  const newRefundTx = new Transaction({ allowUnknownOutputs: true });
  newRefundTx.addInput({
    ...nodeOutPoint,
    sequence,
  });

  const refundPkScript = getP2TRScriptFromPublicKey(receivingPubkey, network);

  newRefundTx.addOutput({
    script: refundPkScript,
    amount: amountSats,
  });
  newRefundTx.addOutput(getEphemeralAnchorOutput());

  return newRefundTx;
}

export function getNextTransactionSequence(
  currSequence?: number,
  forRefresh?: boolean,
): {
  nextSequence: number;
  needRefresh: boolean;
} {
  const currentTimelock = (currSequence || 0) & 0xffff;
  const nextTimelock = currentTimelock - TIME_LOCK_INTERVAL;
  if (forRefresh && nextTimelock <= 100 && currentTimelock > 0) {
    return {
      nextSequence: (1 << 30) | nextTimelock,
      needRefresh: true,
    };
  }

  if (nextTimelock <= 0) {
    throw new Error("timelock interval is less or equal to 0");
  }

  return {
    nextSequence: (1 << 30) | nextTimelock,
    needRefresh: nextTimelock <= 100,
  };
}

export function getEphemeralAnchorOutput(): TransactionOutput {
  return {
    script: new Uint8Array([0x51]), // OP_TRUE
    amount: 0n,
  };
}
